extensions/opencv/MotionDetector.cpp

```diff
@@ -0,0 +1,222 @@
+        ->withAllowableValues<std::string>({"yes", "no"})
+        ->withDefaultValue("no")->build());
+
+core::Relationship MotionDetector::Success(  // NOLINT
+    "success",
+    "Successful to detect motion");
+core::Relationship MotionDetector::Failure(  // NOLINT
+    "failure",
+    "Failure to detect motion");
+
+void MotionDetector::initialize() {
+  std::set<core::Property> properties;
+  properties.insert(ImageEncoding);
+  properties.insert(MinInterestArea);
+  properties.insert(Threshold);
+  properties.insert(BackgroundFrame);
+  properties.insert(DilateIter);
+  properties.insert(PrevAsBackground);
+  setSupportedProperties(std::move(properties));
+
+  std::set<core::Relationship> relationships;
+  relationships.insert(Success);
+  relationships.insert(Failure);
+  setSupportedRelationships(std::move(relationships));
+}
+
+void MotionDetector::onSchedule(core::ProcessContext *context, core::ProcessSessionFactory *sessionFactory) {
+  std::string value;
+
+  if (context->getProperty(ImageEncoding.getName(), value)) {
+    image_encoding_ = value;
+  }
+
+  if (context->getProperty(MinInterestArea.getName(), value)) {
+    core::Property::StringToInt(value, min_area_);
+  }
+
+  if (context->getProperty(Threshold.getName(), value)) {
+    core::Property::StringToInt(value, threshold_);
+  }
+
+  if (context->getProperty(DilateIter.getName(), value)) {
+    core::Property::StringToInt(value, dil_iter_);
+  }
+
+  if (context->getProperty(BackgroundFrame.getName(), value) && !value.empty()) {
+    bg_img_ = cv::imread(value, cv::IMREAD_GRAYSCALE);
+    cv::GaussianBlur(bg_img_, bg_img_, cv::Size(21, 21), 0, 0);
+    bg_img_.convertTo(background_, CV_32F);
+  }
+
+  if (context->getProperty(PrevAsBackground.getName(), value)) {
+    update_bg_ = (value == "yes");
+  }
+
+  logger_->log_trace("MotionDetector processor scheduled");
+}
+
+bool MotionDetector::detectAndDraw(cv::Mat &frame, cv::Mat &diff) {
+  cv::Mat gray;
+  cv::Mat img_diff, thresh;
+  std::vector<cv::Mat> contours;
+
+  cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);
+  cv::GaussianBlur(gray, gray, cv::Size(21, 21), 0, 0);
+
+  // Get difference between current frame and background
+  cv::absdiff(gray, bg_img_, img_diff);
+  cv::threshold(img_diff, thresh, threshold_, 255, cv::THRESH_BINARY);
+  // Image processing.
+  cv::dilate(thresh, thresh, cv::Mat(), cv::Point(-1, -1), dil_iter_);
+  cv::findContours(thresh, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);
+
+  // Finish process
+  thresh.copyTo(diff);
+
+  bool moved = false;
+  for(const auto &contour : contours) {
+    auto area = cv::contourArea(contour);
+    if (area < min_area_) {
+      continue;
+    }
+    // TODO - Nghia: If a contour exists for >n frames, then we should consider it as background
+    // TODO - Nghia: How? Map<rectangle, cnt>?
+    moved = true;
+    cv::Rect bbox = cv::boundingRect(contour);
+    cv::rectangle(frame, bbox.tl(), bbox.br(), cv::Scalar(0, 255, 0), 1, 8, 0);
```

Comment:
Is this responsible for storing all the rectangles that we detect motion in?

Reply:
Yes, it can be looked at that way